File: Heap/239.Sliding_Window_Maximum/Sliding_Window_Maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> q = new LinkedList<>();
        ArrayList<Integer> ret = new ArrayList<>();
        
        for (int i = 0; i < nums.length; i++) {
            while (!q.isEmpty() && nums[i] > nums[q.peekLast()]) {
                q.pollLast();
            }
            q.add(i);
            
            if (i == q.peekFirst() + k) {
                q.pollFirst();
            }
            
            if (i >= k - 1) {
                ret.add(nums[q.peekFirst()]);
            }
        }
        return ret.stream().mapToInt(i->i).toArray();
    }
}

File: Heap/239.Sliding_Window_Maximum/Sliding_Window_Maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # deque, time: O(N), space: O(N), O(k)
        ret = []
        q = collections.deque()  # record index
        
        for i, num in enumerate(nums):
            if q and i == q[0] + k:
                q.popleft()
            
            # remove indices of smaller elements
            while q and num > nums[q[-1]]:
                q.pop()
            q.append(i)

            if i >= k - 1:
                ret.append(nums[q[0]])
        return ret